src/System.Private.CoreLib/shared/System/Marvin.OrdinalIgnoreCase.cs

```diff
@@ -0,0 +1,131 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+using Internal.Runtime.CompilerServices;
+
+#if BIT64
+using nuint = System.UInt64;
+#else
+using nuint = System.UInt32;
+#endif
+
+namespace System
+{
+    internal static partial class Marvin
+    {
+        /// <summary>
+        /// Compute a Marvin OrdinalIgnoreCase hash and collapse it into a 32-bit hash.
+        /// n.b. <paramref name="count"/> is specified as char count, not byte count.
+        /// </summary>
+        public static int ComputeHash32OrdinalIgnoreCase(ref char data, int count, uint p0, uint p1)
+        {
+            nuint ucount = (nuint)count;
+            nuint byteOffset = 0;
+            uint tempValue;
+
+            while (ucount >= 4)
+            {
+                tempValue = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
+                if (!Utf16Utility.AllCharsInUInt32AreAscii(tempValue))
+                {
+                    goto NotAscii;
+                }
+                p0 += Utf16Utility.ConvertAllAsciiCharsInUInt32ToUppercase(tempValue);
+                Block(ref p0, ref p1);
+
+                tempValue = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref data, byteOffset + 4)));
```

Comment:
Have you done the measurements to validate that the manual unrolling is still worth it here with all the extra code? (The comment at https://github.com/dotnet/coreclr/pull/20309#discussion_r223900234 is hidden now.)

Reply:
They're all basically within each other's stddev range.

|                        Method |       Inner loop | StringLength |     Mean |    Error |   StdDev | Scaled |
|------------------------------ |----------------- |------------- |---------:|---------:|---------:|-------:|
| GetHashCode_OrdinalIgnoreCase |  one 32-bit read |         1000 | 871.8 ns | 4.819 ns | 4.272 ns |   1.00 |
| GetHashCode_OrdinalIgnoreCase |  one 64-bit read |         1000 | 874.6 ns | 3.192 ns | 2.986 ns |   1.01 |
| GetHashCode_OrdinalIgnoreCase | two 32-bit reads |         1000 | 868.9 ns | 5.143 ns | 4.559 ns |   1.00 |

I'll simplify the inner loop. This would mean that the logic looks and feels a bit different than the normal Marvin code, but perhaps we're ok with this?